Fix ocaml compilation in ubuntu 22.04 LTS.

kjlau0112 opened the pull request:
Upstream glibc broke ocaml compilation in ubuntu
21 onward. Backport upstream ocaml modification
from commit:17df117b4939486d3285031900587afce5262c8c
to address the build issue and the memory leaks issue
post https://github.com/ocaml/ocaml/pull/10266  modification.

kjlau0112 commented:
[20240206083735.log](https://github.com/intel/meta-intel-sgx/files/14177011/20240206083735.log)
Provided evidence of ocaml-build using ubuntu 22 LTS

AdithyaBaglody commented on recipes-devtools/ocaml/files/0001-Update-ocaml-native-to-fix-upstream-glic-build.patch:
```diff
@@ -0,0 +1,95 @@
+SIGSTKSZ may not be a compile-time constant.
+It is no longer possible to statically allocate the alternate
+signal stack for the main thread.
+
```
please add reference to the upstream PR in this patch. so that we can reconsider this when the changes come in.

kjlau0112 replied:
added, also provide updated build log in ubuntu 22 LTS
[20240219023941.log](https://github.com/intel/meta-intel-sgx/files/14326277/20240219023941.log)

kjlau0112 commented on recipes-devtools/ocaml/files/0001-Update-ocaml-native-to-fix-upstream-glic-build.patch:
```diff
@@ -0,0 +1,96 @@
+signal stack for the main thread.
+
+Upstream-Status:Backport
+Reference: https://github.com/ocaml/ocaml/pull/10266
```
@AdithyaBaglody  added for your reference.

pchand20 commented on recipes-devtools/ocaml/files/0001-Update-ocaml-native-to-fix-upstream-glic-build.patch:
why we removed this signal thread check here ?

kjlau0112 replied:
> why we removed this signal thread check here ?

This changes restore as per https://github.com/ocaml/ocaml/commit/17df117b4939486d3285031900587afce5262c8c
This PR patch also updated  to align with above commit.

pchand20 commented on recipes-devtools/ocaml/files/0001-Update-ocaml-native-to-fix-upstream-glic-build.patch:
where are we freeing this memory?

kjlau0112 replied:
During the early fixing of https://github.com/ocaml/ocaml/pull/10266 , author intended to risk memory leak as per comment below:
https://github.com/ocaml/ocaml/issues/10250#issuecomment-789108059

However after i checked https://github.com/ocaml/ocaml/commits/4.08/ , the glibc and memory leaks issue is further addressed by this commit: https://github.com/ocaml/ocaml/commit/17df117b4939486d3285031900587afce5262c8c and as per commit message alternate signal stack shall free up the memory post main OCaml  thread stops.

kjlau0112 replied:
> where are we freeing this memory?

During the early fixing of https://github.com/ocaml/ocaml/pull/10266 , author intended to risk memory leak as per comment below:
https://github.com/ocaml/ocaml/issues/10250#issuecomment-789108059

However after i checked https://github.com/ocaml/ocaml/commits/4.08/ , the glibc and memory leaks issue is further addressed by this commit: https://github.com/ocaml/ocaml/commit/17df117b4939486d3285031900587afce5262c8c and as per commit message alternate signal stack shall free up the memory post main OCaml thread stops.

pchand20 commented on recipes-devtools/ocaml/files/0001-Update-ocaml-native-to-fix-upstream-glic-build.patch:
above review comment 1 we have removed the check and from here we are calling sending signal , seems we need the previous check based on this event.

kjlau0112 replied:
> above review comment 1 we have removed the check and from here we are calling sending signal , seems we need the previous check based on this event.

The latest changes have removed this  changes, let re-align on  https://github.com/ocaml/ocaml/commit/17df117b4939486d3285031900587afce5262c8c changes

kjlau0112 committed:
Fix ocaml compilation in ubuntu 22.04 LTS.

Upstream glibc broke ocaml compilation in ubuntu
21 onward. Backport upstream ocaml modification
from commit:17df117b4939486d3285031900587afce5262c8c
to address the build issue and the memory leaks issue
post ocaml/ocaml#10266  modification.

Reference:https://github.com/ocaml/ocaml/commit/17df117b4939486d3285031900587afce5262c8c

Signed-off-by: Karn Jye Lau <karn.jye.lau@intel.com>

kjlau0112 commented:
[20240222045957.log](https://github.com/intel/meta-intel-sgx/files/14369673/20240222045957.log)
Ubuntu 22 LTS build log per https://github.com/kjlau0112/meta-intel-sgx/commit/6b7a1ffb2bf9dcda327b7e1dcc1bc9d5fac742c6 commit.

pchand20 reviewed (approved):
LGTM

kartikeyparmar commented:
@pchand20 Can you please merge this PR as review from your end completed?
